src/Control/Effect/State.hs

```diff
@@ -119,6 +129,57 @@ instance (Carrier sig m, Effect sig) => Carrier (State s :+: sig) (StateC s m) w
   eff (L (Put s k)) = StateC (\ _ -> runState s k)
   eff (R other)     = StateC (\ s -> eff (handle (s, ()) (uncurry runState) other))
 
+newtype LazyStateC s m a = LazyStateC { runLazyStateC :: s -> m (s, a) }
+
+instance Functor m => Functor (LazyStateC s m) where
+  fmap f m = LazyStateC $ \ s -> fmap (\ ~(s', a) -> (s', f a)) $ runLazyStateC m s
+
+instance (Functor m, Monad m) => Applicative (LazyStateC s m) where
+  pure a = LazyStateC $ \ s -> pure (s, a)
+  LazyStateC mf <*> LazyStateC mx = LazyStateC $ \ s -> do
+    ~(s', f) <- mf s
+    ~(s'', x) <- mx s'
+    return (s'', f x)
+
+instance Monad m => Monad (LazyStateC s m) where
+  m >>= k  = LazyStateC $ \ s -> do
+    ~(s', a) <- runLazyStateC m s
+    runLazyStateC (k a) s'
+
+instance (Alternative m, Monad m) => Alternative (LazyStateC s m) where
+  empty = LazyStateC (const empty)
+  LazyStateC l <|> LazyStateC r = LazyStateC (\ s -> l s <|> r s)
+
+instance MonadIO m => MonadIO (LazyStateC s m) where
+  liftIO io = LazyStateC (\ s -> (,) s <$> liftIO io)
+
+instance (Alternative m, Monad m) => MonadPlus (LazyStateC s m)
+
+instance (Carrier sig m, Effect sig) => Carrier (State s :+: sig) (LazyStateC s m) where
+  eff (L (Get   k)) = LazyStateC (\ s -> runLazyState s (k s))
+  eff (L (Put s k)) = LazyStateC (\ _ -> runLazyState s k)
+  eff (R other)     = LazyStateC (\ s -> eff (handle (s, ()) (uncurry runLazyState) other))
+
+-- | Run a lazy 'State' effect, yielding the result value and the final state.
+--   More programs terminate with lazy state than strict state, but injudicious
+--   use of lazy state may lead to thunk buildup.
+--
+--   prop> run (evalLazyState a (pure b)) == b
+--   prop> take 5 . snd . run $ runLazyState (traverse pure [1..]) == [1,2,3,4,5]
```

Comment:
I’m pretty certain this doctest isn’t being run, because:

- the relative priority of `$` and `==` would mean we’re equating `VoidC Integer` and then `run`ning `Bool`
- we’re not giving `runLazyState` an initial state value.

Reply:
I double-checked that we do in fact infinite loop if we try this with our strict `StateC` 👍